packages/effect/src/internal/rcMap.ts

```diff
@@ -0,0 +1,194 @@
+      options.lookup as any,
+      context,
+      scope,
+      options.idleTimeToLive ? Duration.decode(options.idleTimeToLive) : Duration.zero
+    )
+    return core.as(
+      scope.addFinalizer(() =>
+        core.suspend(() => {
+          if (self.state._tag === "Closed") {
+            return core.void
+          }
+          const map = self.state.map
+          self.state = { _tag: "Closed" }
+          return core.forEachSequentialDiscard(
+            map,
+            ([, entry]) => core.scopeClose(entry.scope, core.exitVoid)
+          ).pipe(
+            core.tap(() => {
+              MutableHashMap.clear(map)
+            }),
+            self.semaphore.withPermits(1)
+          )
+        })
+      ),
+      self
+    )
+  })
+
+/** @internal */
+export const get: {
+  <K>(key: K): <A, E>(self: RcMap.RcMap<K, A, E>) => Effect<A, E, Scope.Scope>
+  <K, A, E>(self: RcMap.RcMap<K, A, E>, key: K): Effect<A, E, Scope.Scope>
+} = dual(
+  2,
+  <K, A, E>(self_: RcMap.RcMap<K, A, E>, key: K): Effect<A, E, Scope.Scope> => {
+    const self = self_ as RcMapImpl<K, A, E>
+    return core.uninterruptibleMask((restore) =>
+      core.suspend(() => {
+        if (self.state._tag === "Closed") {
+          return core.interrupt
+        }
+        const state = self.state
+        const o = MutableHashMap.get(state.map, key)
+        if (o._tag === "Some") {
+          const entry = o.value
+          entry.refCount++
+          return entry.fiber
+            ? core.as(core.interruptFiber(entry.fiber), entry)
+            : core.succeed(entry)
+        }
+        const acquire = self.lookup(key)
+        return fiberRuntime.scopeMake().pipe(
+          coreEffect.bindTo("scope"),
+          coreEffect.bind("deferred", () => core.deferredMake<A, E>()),
+          core.tap(({ deferred, scope }) =>
+            restore(core.fiberRefLocally(
+              acquire as Effect<A, E>,
+              core.currentContext,
+              Context.add(self.context, fiberRuntime.scopeTag, scope)
+            )).pipe(
+              core.exit,
+              core.flatMap((exit) => core.deferredDone(deferred, exit)),
+              circular.forkIn(scope)
+            )
+          ),
+          core.map(({ deferred, scope }) => {
+            const entry: State.Entry<A, E> = {
+              deferred,
+              scope,
+              fiber: undefined,
+              refCount: 1
+            }
+            MutableHashMap.set(state.map, key, entry)
+            return entry
+          })
+        )
+      }).pipe(
+        self.semaphore.withPermits(1),
+        coreEffect.bindTo("entry"),
+        coreEffect.bind("scope", () => fiberRuntime.scopeTag),
+        core.tap(({ entry, scope }) =>
+          scope.addFinalizer(() =>
+            core.suspend(() => {
+              entry.refCount--
+              if (entry.refCount > 0) {
+                return core.void
+              } else if (Duration.isZero(self.idleTimeToLive)) {
```

Comment:
Actually, when i use this, i only use zero-duration to release resource at the next event loop tick.
I think it's better to check if `idleTimeToLive` is undefined, because `undefined` and '0 millis' is not the same